Mathlib/Computability/NFA.lean

```diff
@@ -407,6 +404,112 @@ theorem mem_accepts_reverse {x : List α} : x ∈ M.reverse.accepts ↔ x.revers
 
 end NFA
 
+namespace NFA
+
+section concat
+
+open Sum
+
+/-!
+### Declarations about `NFA.concat`
+
+We provide a direct construction of concatenated NFAs without ε-transitions: if `M1` accepts
+language `L1` and `M2` accepts language `L2` then `M1.concat M2` accepts language `L1 * L2`.
+In order to construct `M1.concat M2`, we must create transitions from accepting states in `M1`
+to states immediatately reachable after a single transition from states in `M2.start`. Furthermore,
+we must include states of `M1.accept` in `(M1.concat M2).accept` if `[] ∈ M2.accepts`.
+
+Traditionally, concatenation is constructed for `εNFA`, not plain `NFA`, by adding ε-transitions
+from states in `M1.accept` to `M2.start`. Since `NFA` does not include ε-transitions, we must
+include the aforementioned transitions for `NFA.concatStep`:
+`⋃ s2 ∈ M2.start, { s2' ∈ M2.step s2 a | s1 ∈ M1.accept }`.
+
+This construction for unweighted `NFA` provides insight into the weighted case, where the weighted
+analogue for `NFA` is different from the weighted analogue for `εNFA`.
+-/
+
+variable {σ1 σ2 : Type v} (M1 : NFA α σ1) (M2 : NFA α σ2)
+
+/-- `M1.concat M2` is the concatenation of `M1` and `M2` achieved without ε-transitions. -/
+@[simps]
+def concat : NFA α (σ1 ⊕ σ2) where
+  step
+  | .inl s1, a =>
+    inl '' M1.step s1 a ∪ inr '' ⋃ s2 ∈ M2.start, { s2' ∈ M2.step s2 a | s1 ∈ M1.accept }
+  | .inr s2, a => inr '' M2.step s2 a
+  start := inl '' M1.start
+  accept := inl '' { s1 ∈ M1.accept | [] ∈ M2.accepts } ∪ inr '' M2.accept
+
+lemma stepSet_concat_inr {S2 : Set σ2} {a : α} :
+    (M1.concat M2).stepSet (inr '' S2) a = inr '' M2.stepSet S2 a := by
+  ext (s1 | s2) <;> simp [stepSet]
+
+lemma stepSet_concat_inl {S1 : Set σ1} {a : α} :
+    (M1.concat M2).stepSet (inl '' S1) a =
+      inl '' M1.stepSet S1 a ∪
+      inr '' ⋃ s1 ∈ S1, ⋃ s2 ∈ M2.start, { s2' ∈ M2.step s2 a | s1 ∈ M1.accept } := by
+  ext (s1 | s2) <;> simp [stepSet]
+
+lemma acceptsFrom_concat_inr {S2 : Set σ2} :
+    (M1.concat M2).acceptsFrom (inr '' S2) = M2.acceptsFrom S2 := by
+  ext y
+  induction y generalizing S2 with
+  | nil => simp
+  | cons a y ih => simp [←ih, stepSet_concat_inr]
+
+theorem acceptsFrom_concat_inl {S1 : Set σ1} :
+    (M1.concat M2).acceptsFrom (inl '' S1) = M1.acceptsFrom S1 * M2.accepts := by
+  ext z
+  rw [Language.mul_def, Set.mem_image2]
+  induction z generalizing S1 with
+  | nil =>
+    suffices h : (∃ a ∈ S1, a ∈ M1.accept ∧ [] ∈ M2.accepts)
+      ↔ (∃ s ∈ S1, s ∈ M1.accept) ∧ [] ∈ M2.accepts by simpa [nil_mem_acceptsFrom M1]
+    tauto
+  | cons a z ih =>
+    simp only [cons_mem_acceptsFrom, ↓stepSet_concat_inl, stepSet, acceptsFrom_union,
+      acceptsFrom_concat_inr, acceptsFrom_iUnion, add_eq_sup, max, SemilatticeSup.sup,
```

Comment:
Why do you need to unfold `max` and `SemilatticeSup.sup`? Are we missing a lemma here?

Reply:
Ah, thanks, good catch! I could replace these with `Language.add_def`.